File: apps/web/lib/rate-limit.ts
/**
 * Minimal in-memory fixed-window rate limiter (Faz 8 hardening). Single-node
 * scope — matches the app's single-process authoritative model (invariant #4);
 * a horizontally-scaled deployment would swap this for Redis. Used to throttle
 * auth attempts (brute-force / credential-stuffing defense) keyed by IP and email.
 */
interface Window {
  count: number;
  resetAt: number;
}

const windows = new Map<string, Window>();

const SWEEP_THRESHOLD = 5000; // start pruning only when map is large — sweep itself is O(n)

/** Prune expired windows opportunistically so the map can't grow unbounded. */
function sweep(now: number): void {
  if (windows.size < SWEEP_THRESHOLD) return;
  for (const [key, w] of windows) if (now >= w.resetAt) windows.delete(key);
}

export interface RateResult {
  ok: boolean;
  /** Seconds until the window resets (only meaningful when !ok). */
  retryAfter: number;
}

/**
 * Consume one unit against `key`. Returns ok=false once `limit` is reached within
 * `windowMs`; the window resets `windowMs` after the first hit.
 */
export function rateLimit(key: string, limit: number, windowMs: number): RateResult {
  const now = Date.now();
  sweep(now);
  const w = windows.get(key);
  if (!w || now >= w.resetAt) {
    windows.set(key, { count: 1, resetAt: now + windowMs });
    return { ok: true, retryAfter: 0 };
  }
  if (w.count >= limit) {
    return { ok: false, retryAfter: Math.ceil((w.resetAt - now) / 1000) };
  }
  w.count += 1;
  return { ok: true, retryAfter: 0 };
}

/** Test/util: clear all windows. */
export function resetRateLimits(): void {
  windows.clear();
}

File: services/api-go/internal/rest/ratelimit.go
package rest

import (
	"sync"
	"time"
)

// RateLimiter is an in-memory fixed-window limiter (rate-limit.ts parity).
// Single-node scope, matching the single-process model (invariant #4);
// keyed by IP and email for auth throttles.

type window struct {
	count   int
	resetAt time.Time
}

// RateLimiter guards keys against bursts. Zero value is usable (but shares no
// state until created with NewRateLimiter — prefer the constructor).
type RateLimiter struct {
	mu      sync.Mutex
	windows map[string]window
}

// sweepThreshold mirrors SWEEP_THRESHOLD (prune only when large).
const sweepThreshold = 5000

// NewRateLimiter creates an empty limiter.
func NewRateLimiter() *RateLimiter {
	return &RateLimiter{windows: make(map[string]window)}
}

// Allow consumes one unit: false once limit is reached within window.
// retryAfter rounds up the seconds until reset (0 when allowed).
func (l *RateLimiter) Allow(key string, limit int, windowMs int64) (ok bool, retryAfter int64) {
	now := time.Now()
	l.mu.Lock()
	defer l.mu.Unlock()
	if len(l.windows) >= sweepThreshold {
		for k, w := range l.windows {
			if !now.Before(w.resetAt) {
				delete(l.windows, k)
			}
		}
	}
	w, found := l.windows[key]
	if !found || !now.Before(w.resetAt) {
		l.windows[key] = window{count: 1, resetAt: now.Add(time.Duration(windowMs) * time.Millisecond)}
		return true, 0
	}
	if w.count >= limit {
		return false, int64((w.resetAt.Sub(now) + time.Second - 1) / time.Second)
	}
	w.count++
	l.windows[key] = w
	return true, 0
}

// Reset clears all windows (tests).
func (l *RateLimiter) Reset() {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.windows = make(map[string]window)
}
